dm-manager/pkg/device/device.go

```
@@ -0,0 +1,402 @@
+		statusv1.StatusIndication_STATUS_INDICATION_IN_PROGRESS,
+	)
+	if err != nil {
+		log.Err(err).Msgf("failed to update device info")
+		return request.Fail(err), statusv1.StatusIndication_STATUS_INDICATION_ERROR, err
+	}
+
+	// assumption is that we don't have to check if power actions are supported, as AMT supports it since 1.0.0
+	// https://en.wikipedia.org/wiki/Intel_AMT_versions
+	powerAction, err := dc.MpsClient.PostApiV1AmtPowerActionGuidWithResponse(ctx, request.ID.GetHostUUID(),
+		mps.PostApiV1AmtPowerActionGuidJSONRequestBody{
+			Action: powerMapping[invHost.GetDesiredPowerState()],
+		})
+	if err != nil {
+		log.Err(err).Msgf("failed to send power action to MPS")
+		return request.Fail(err), statusv1.StatusIndication_STATUS_INDICATION_ERROR, err
+	}
+
+	log.Debug().Msgf("power action response for %v with status code %v - %v", request.ID.GetHostUUID(),
+		powerAction.HTTPResponse, string(powerAction.Body))
+
+	if powerAction.StatusCode() != http.StatusOK {
+		err = errors.Errorf("%v", string(powerAction.Body))
+		log.Err(err).
+			Msgf("expected to get 2XX, but got %v", powerAction.StatusCode())
+		return request.Fail(err), statusv1.StatusIndication_STATUS_INDICATION_ERROR, err
+	}
+
+	// intentionally comparing whole body, as there are cases where MPS defines variable as lowercase
+	// but it is uppercase instead
+	if !strings.EqualFold(*powerAction.JSON200.Body.ReturnValueStr, "SUCCESS") {
+		log.Err(errors.Errorf("power request sent successfully, but received unexpected response")).
+			Msgf("expected to receive SUCCESS, but got %s", string(powerAction.Body))
+
+		// NOT_READY is returned when device is powered off but someone tries to reboot it
+		// due to incorrect start state. Most of actions require POWER_ON as start state.
+		// https://device-management-toolkit.github.io/docs/2.27/Reference/powerstates/
+		if strings.EqualFold(*powerAction.JSON200.Body.ReturnValueStr, "NOT_READY") {
+			return dc.forcePowerOn(ctx, request)
```

do we really want to do this ? Shall we just report an error ?

---

I belive we should let the user correct the issue (so we should just return an error). On friday we discussed this approach because there was the loop issue.